// @s2-dev/streamstore/dist/cjs/accessTokens.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.S2AccessTokens = void 0;
const error_js_1 = require("./error.js");
const index_js_1 = require("./generated/index.js");
class S2AccessTokens {
    client;
    constructor(client) {
        this.client = client;
    }
    /**
     * List access tokens.
     *
     * @param args.prefix Filter to IDs beginning with this prefix
     * @param args.start_after Filter to IDs lexicographically after this value
     * @param args.limit Max results (up to 1000)
     */
    async list(args, options) {
        const response = await (0, index_js_1.listAccessTokens)({
            client: this.client,
            query: args,
            ...options,
        });
        if (response.error) {
            throw new error_js_1.S2Error({
                message: response.error.message,
                code: response.error.code ?? undefined,
                status: response.response.status,
            });
        }
        return response.data;
    }
    /**
     * Issue a new access token.
     *
     * @param args.id Unique token ID (1-96 bytes)
     * @param args.scope Token scope (operations and resource sets)
     * @param args.auto_prefix_streams Namespace stream names by configured prefix scope
     * @param args.expires_at Expiration in ISO 8601; defaults to requestor's token expiry
     */
    async issue(args, options) {
        const response = await (0, index_js_1.issueAccessToken)({
            client: this.client,
            body: args,
            ...options,
        });
        if (response.error) {
            throw new error_js_1.S2Error({
                message: response.error.message,
                code: response.error.code ?? undefined,
                status: response.response.status,
            });
        }
        return response.data;
    }
    /**
     * Revoke an access token by ID.
     *
     * @param args.id Token ID to revoke
     */
    async revoke(args, options) {
        const response = await (0, index_js_1.revokeAccessToken)({
            client: this.client,
            path: args,
            ...options,
        });
        if (response.error) {
            throw new error_js_1.S2Error({
                message: response.error.message,
                code: response.error.code ?? undefined,
                status: response.response.status,
            });
        }
        return response.data;
    }
}
exports.S2AccessTokens = S2AccessTokens;
