// @s2-dev/streamstore/dist/cjs/metrics.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.S2Metrics = void 0;
const error_js_1 = require("./error.js");
const index_js_1 = require("./generated/index.js");
class S2Metrics {
    client;
    constructor(client) {
        this.client = client;
    }
    /**
     * Account-level metrics.
     *
     * @param args.set Metric set to return
     * @param args.start Optional start timestamp (Unix seconds)
     * @param args.end Optional end timestamp (Unix seconds)
     * @param args.interval Optional aggregation interval for timeseries sets
     */
    async account(args, options) {
        const response = await (0, index_js_1.accountMetrics)({
            client: this.client,
            query: args,
            ...options,
        });
        if (response.error) {
            throw new error_js_1.S2Error({
                message: response.error.message,
                code: response.error.code ?? undefined,
                status: response.response.status,
            });
        }
        return response.data;
    }
    /**
     * Basin-level metrics.
     *
     * @param args.basin Basin name
     * @param args.set Metric set to return
     * @param args.start Optional start timestamp (Unix seconds)
     * @param args.end Optional end timestamp (Unix seconds)
     * @param args.interval Optional aggregation interval for timeseries sets
     */
    async basin(args, options) {
        const response = await (0, index_js_1.basinMetrics)({
            client: this.client,
            path: args,
            query: args,
            ...options,
        });
        if (response.error) {
            throw new error_js_1.S2Error({
                message: response.error.message,
                code: response.error.code ?? undefined,
                status: response.response.status,
            });
        }
        return response.data;
    }
    /**
     * Stream-level metrics.
     *
     * @param args.basin Basin name
     * @param args.stream Stream name
     * @param args.set Metric set to return
     * @param args.start Optional start timestamp (Unix seconds)
     * @param args.end Optional end timestamp (Unix seconds)
     * @param args.interval Optional aggregation interval for timeseries sets
     */
    async stream(args, options) {
        const response = await (0, index_js_1.streamMetrics)({
            client: this.client,
            path: args,
            query: args,
            ...options,
        });
        if (response.error) {
            throw new error_js_1.S2Error({
                message: response.error.message,
                code: response.error.code ?? undefined,
                status: response.response.status,
            });
        }
        return response.data;
    }
}
exports.S2Metrics = S2Metrics;
